tokenserver/views.py

```diff
@@ -349,12 +374,22 @@ def return_token(request):
             raise _unauthorized('new-users-disabled')
         with metrics_timer('tokenserver.backend.allocate_user', request):
             user = backend.allocate_user(service, email, generation,
-                                         client_state)
+                                         client_state,
+                                         keys_changed_at=keys_changed_at)
 
     # Update if this client is ahead of previously-seen clients.
     updates = {}
     if generation > user['generation']:
         updates['generation'] = generation
+    if keys_changed_at > user['keys_changed_at']:
+        # TODO: If there's a generation number available, then a change
+        # in keys should correspond to a change in generation number.
+        # We can't accurately detect this case until all servers are reliably
+        # writing keys_changed_at to the database, so it will need to wait
+        # and be deployed separately.
+        # if generation > 0 and 'generation' not in updates:
```

Comment:
This looks sound, my only thought is if this is in fact the only valid case:

>     * It detects that their `keys_changed_at` value has increased, but that `generation` matches what is currently stored in the db.

then could this be stricter?

```suggestion
        if generation > 0 and 'generation' not in updates and generation != user['generation']:
```

but with this being a hopefully temporary situation, removing this check entirely is fine.

Reply:
Good point, yes, I think you're right!

Reply:
In trying to add a test for this, I realized such auth attempts will already be rejected by the "Error out if this client provided a generation number, but it is behind that of some previously-seen client" logic further down the file. I'm going to keep it because it'll be some nice extra check against a misbehaving FxA server.